### PR TITLE
Extended sets

### PR DESCRIPTION
**Pull request by PierreBoyeau**

### CODE REVIEW

File: src/scvi_v2/_utils.py [hunk 1]
```diff
@@ -221,3 +224,166 @@ def permute_compute(w, x, key):
...
+        X=mylfcs.values,
+        obs=myadata.obs,
+        var=myadata.var,
+    )
+
+    # estimate gene clusters
+    gene_include_rule = np.logical_and(myadata.X.mean(0) > 0.01, lfc_ad.X.std(0) >= 0.1)
+    lfc_ad_transpose = lfc_ad[:, gene_include_rule].copy().T
+    sc.pp.neighbors(lfc_ad_transpose, n_neighbors=15, use_rep="X", metric="cosine")
+    sc.tl.umap(lfc_ad_transpose, min_dist=0.1)
```
Comment (canergen) [inline src/scvi_v2/_utils.py]: Move this to line 285.

File: src/scvi_v2/_utils.py [hunk 2]
```diff
@@ -221,3 +224,166 @@ def permute_compute(w, x, key):
...
+    lfc_ad = sc.AnnData(
+        X=mylfcs.values,
+        obs=myadata.obs,
+        var=myadata.var,
+    )
+
+    # estimate gene clusters
+    gene_include_rule = np.logical_and(myadata.X.mean(0) > 0.01, lfc_ad.X.std(0) >= 0.1)
+    lfc_ad_transpose = lfc_ad[:, gene_include_rule].copy().T
+    sc.pp.neighbors(lfc_ad_transpose, n_neighbors=15, use_rep="X", metric="cosine")
```
Comment (canergen) [inline src/scvi_v2/_utils.py]: Expose parameters as kwargs.

File: src/scvi_v2/_utils.py [hunk 3]
```diff
@@ -221,3 +224,166 @@ def permute_compute(w, x, key):
...
+    sc.tl.umap(lfc_ad_transpose, min_dist=0.1)
+    sc.tl.leiden(lfc_ad_transpose, resolution=0.4, key_added="gene_leiden")
+    if plot:
+        sc.pl.umap(lfc_ad_transpose, color="gene_leiden")
+
+    # assign gene clusters
+    myadata.var.loc[:, "gene_cluster"] = "NA"
```
Comment (canergen) [inline src/scvi_v2/_utils.py:297]: myadata.var["gene_cluster"]

File: src/scvi_v2/_utils.py [hunk 4]
```diff
@@ -221,3 +224,166 @@ def permute_compute(w, x, key):
...
+    myadata.var.loc[lfc_ad_transpose.obs.index, "gene_cluster"] = lfc_ad_transpose.obs["gene_leiden"].astype(str).values
+
+    # compute cell scores
+    cell_score_keys = []
+    for g_cluster in myadata.var.loc[:, "gene_cluster"].unique():
+        if g_cluster == "NA":
+            continue
+        sub_ad = lfc_ad[:, myadata.var["gene_cluster"] == g_cluster].X
+        signs_ = np.sum(np.sign(sub_ad), 1) / sub_ad.shape[-1]
```
Comment (canergen) [inline src/scvi_v2/_utils.py:306]: This is scoring just the signs without looking at the LFC size?

File: src/scvi_v2/_utils.py [hunk 5]
```diff
@@ -221,3 +224,166 @@ def permute_compute(w, x, key):
...
+
+        if score_quantile is not None:
+            score_threshold = np.quantile(scores, score_quantile)
+        is_in_cluster = scores >= score_threshold
+        myadata.obs.loc[:, cell_cluster_key] = is_in_cluster
+        keys_to_plot.append(cell_cluster_key)
+
+        is_in_extended_cluster = is_in_cluster.copy()
+        for _ in range(extend_nhops):
+            neighbors = myadata.obsp["connectivities"][is_in_cluster, :].toarray()
```
Comment (canergen) [inline src/scvi_v2/_utils.py:335]: The initial idea was to extend by nearest neighbors to counterfactuals? Is there a reason you updated to u space?

**Review by canergen — COMMENTED** [review]
I don't think we should include the definition of cell_sets in the model.py code. My idea was to add the code to extend sets in zspace based on counterfactuals and run pseudobulk on those. The current way is more of a reproducibility script as this is quite bulky. So scratch starts with a cell set, extends it with cells in zspace closest to counterfactual and run pseudobulk.

File: src/scvi_v2/_utils.py (hunk 4 above)
Comment (PierreBoyeau) [reply src/scvi_v2/_utils.py:306]: Yes, I did not understand the exact rule you used in the notebooks. What rule did you use exactly?

File: src/scvi_v2/_utils.py (hunk 5 above)
Comment (PierreBoyeau) [reply src/scvi_v2/_utils.py:335]: I am not sure I understand how the procedure in z space would look like. The intuition behind doing it in u space is to extend the considered cell population to its immediate neighbors.
Let's talk about it online!

**Pull request closed without merging**